File: packages/domain/src/confidence-engine/calibration.ts
/**
 * Calibration dataset export for later model calibration and BB-047 evaluation.
 * Exports only deterministic, versioned confidence records with optional observed outcomes.
 */
import type { ClaimClass } from '@black-book/schemas';
import type { AuditedConfidenceResult } from './engine.js';

export const CONFIDENCE_CALIBRATION_DATASET_VERSION =
  'confidence-calibration-dataset.v1' as const;

export type ConfidenceCalibrationCase = {
  readonly claimId: string;
  readonly claimVersionId: string;
  readonly claimClass: ClaimClass;
  readonly confidence: AuditedConfidenceResult;
  readonly observedOutcome?: boolean;
  readonly outcomeSource?: string;
};

export type ConfidenceCalibrationRow = {
  readonly claimId: string;
  readonly claimVersionId: string;
  readonly claimClass: ClaimClass;
  readonly score: number;
  readonly threshold: number;
  readonly passesPublishThreshold: boolean;
  readonly components: AuditedConfidenceResult['components'];
  readonly policyVersion: string;
  readonly engineVersion: string;
  readonly componentVersions: AuditedConfidenceResult['audit']['componentVersions'];
  readonly independentLineageCount: number;
  readonly supportingEvidenceCount: number;
  readonly contradictingEvidenceCount: number;
  readonly inputFingerprints: AuditedConfidenceResult['audit']['inputFingerprints'];
  readonly calculatedAt: string;
  readonly observedOutcome?: boolean;
  readonly outcomeSource?: string;
};

export type ConfidenceCalibrationDataset = {
  readonly schemaVersion: typeof CONFIDENCE_CALIBRATION_DATASET_VERSION;
  readonly exportedAt: string;
  readonly rows: readonly ConfidenceCalibrationRow[];
};

/** Export stable row order so the same cases and timestamp serialize identically. */
export function exportConfidenceCalibrationDataset(input: {
  readonly cases: readonly ConfidenceCalibrationCase[];
  readonly exportedAt: string;
}): ConfidenceCalibrationDataset {
  return {
    schemaVersion: CONFIDENCE_CALIBRATION_DATASET_VERSION,
    exportedAt: input.exportedAt,
    rows: [...input.cases]
      .sort(
        (left, right) =>
          left.claimId.localeCompare(right.claimId) ||
          left.claimVersionId.localeCompare(right.claimVersionId),
      )
      .map(({ claimId, claimVersionId, claimClass, confidence, observedOutcome, outcomeSource }) => ({
        claimId,
        claimVersionId,
        claimClass,
        score: confidence.score,
        threshold: confidence.threshold,
        passesPublishThreshold: confidence.passesPublishThreshold,
        components: confidence.components,
        policyVersion: confidence.policyVersion,
        engineVersion: confidence.audit.engineVersion,
        componentVersions: confidence.audit.componentVersions,
        independentLineageCount: confidence.independentLineageCount,
        supportingEvidenceCount: confidence.supportingEvidenceCount,
        contradictingEvidenceCount: confidence.contradictingEvidenceCount,
        inputFingerprints: confidence.audit.inputFingerprints,
        calculatedAt: confidence.calculatedAt,
        ...(observedOutcome !== undefined ? { observedOutcome } : {}),
        ...(outcomeSource !== undefined ? { outcomeSource } : {}),
      })),
  };
}

File: workers/research/src/black_book_research/confidence_engine/calibration.py
"""Schema-compatible calibration dataset export for confidence assessments."""

from __future__ import annotations

from dataclasses import dataclass

from .models import ConfidenceResult

CONFIDENCE_CALIBRATION_DATASET_VERSION = "confidence-calibration-dataset.v1"


@dataclass(frozen=True, slots=True)
class ConfidenceCalibrationCase:
    claim_id: str
    claim_version_id: str
    claim_class: str
    confidence: ConfidenceResult
    observed_outcome: bool | None = None
    outcome_source: str | None = None


def _components(result: ConfidenceResult) -> dict[str, float]:
    value = result.components
    return {
        "sourceAuthority": value.source_authority,
        "directness": value.directness,
        "lineageIndependence": value.lineage_independence,
        "temporalProximity": value.temporal_proximity,
        "geographicPrecision": value.geographic_precision,
        "entityMatchQuality": value.entity_match_quality,
        "extractionQuality": value.extraction_quality,
        "contradictionPenalty": value.contradiction_penalty,
    }


def export_confidence_calibration_dataset(
    cases: tuple[ConfidenceCalibrationCase, ...],
    *,
    exported_at: str,
) -> dict[str, object]:
    """Export deterministic camel-case rows accepted by the shared JSON Schema."""
    rows: list[dict[str, object]] = []
    for case in sorted(
        cases, key=lambda value: (value.claim_id, value.claim_version_id)
    ):
        result = case.confidence
        row: dict[str, object] = {
            "claimId": case.claim_id,
            "claimVersionId": case.claim_version_id,
            "claimClass": case.claim_class,
            "score": result.score,
            "threshold": result.threshold,
            "passesPublishThreshold": result.passes_publish_threshold,
            "components": _components(result),
            "policyVersion": result.policy_version,
            "engineVersion": result.audit.engine_version,
            "componentVersions": result.audit.component_versions,
            "independentLineageCount": result.independent_lineage_count,
            "supportingEvidenceCount": result.supporting_evidence_count,
            "contradictingEvidenceCount": result.contradicting_evidence_count,
            "inputFingerprints": result.audit.input_fingerprints,
            "calculatedAt": result.calculated_at,
        }
        if case.observed_outcome is not None:
            row["observedOutcome"] = case.observed_outcome
        if case.outcome_source is not None:
            row["outcomeSource"] = case.outcome_source
        rows.append(row)
    return {
        "schemaVersion": CONFIDENCE_CALIBRATION_DATASET_VERSION,
        "exportedAt": exported_at,
        "rows": rows,
    }
